iree/hal/local/elf/arch/x86_64.c

```diff
@@ -0,0 +1,200 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/base/target_platform.h"
+#include "iree/hal/local/elf/arch.h"
+
+#if defined(IREE_ARCH_X86_64)
+
+//==============================================================================
+// ELF machine type/ABI
+//==============================================================================
+
+bool iree_elf_arch_is_valid(const iree_elf_ehdr_t* ehdr) {
+  return ehdr->e_machine == 0x3E;  // EM_X86_64 / 62
+}
+
+//==============================================================================
+// ELF relocations
+//==============================================================================
+
+enum {
+  IREE_ELF_R_X86_64_NONE = 0,       // No reloc
+  IREE_ELF_R_X86_64_64 = 1,         // Direct 64 bit
+  IREE_ELF_R_X86_64_PC32 = 2,       // PC relative 32 bit signed
+  IREE_ELF_R_X86_64_GOT32 = 3,      // 32 bit GOT entry
+  IREE_ELF_R_X86_64_PLT32 = 4,      // 32 bit PLT address
+  IREE_ELF_R_X86_64_COPY = 5,       // Copy symbol at runtime
+  IREE_ELF_R_X86_64_GLOB_DAT = 6,   // Create GOT entry
+  IREE_ELF_R_X86_64_JUMP_SLOT = 7,  // Create PLT entry
+  IREE_ELF_R_X86_64_RELATIVE = 8,   // Adjust by program base
+  IREE_ELF_R_X86_64_GOTPCREL = 9,   // 32 bit signed pc relative offset to GOT
+  IREE_ELF_R_X86_64_32 = 10,        // Direct 32 bit zero extended
+  IREE_ELF_R_X86_64_32S = 11,       // Direct 32 bit sign extended
+  IREE_ELF_R_X86_64_16 = 12,        // Direct 16 bit zero extended
+  IREE_ELF_R_X86_64_PC16 = 13,      // 16 bit sign extended pc relative
+  IREE_ELF_R_X86_64_8 = 14,         // Direct 8 bit sign extended
+  IREE_ELF_R_X86_64_PC8 = 15,       // 8 bit sign extended pc relative
+  IREE_ELF_R_X86_64_PC64 = 24,      // Place relative 64-bit signed
+};
+
+static iree_status_t iree_elf_arch_x86_64_apply_rela(
+    iree_elf_relocation_state_t* state, iree_host_size_t rela_count,
+    const iree_elf_rela_t* rela_table) {
+  for (iree_host_size_t i = 0; i < rela_count; ++i) {
+    const iree_elf_rela_t* rela = &rela_table[i];
+    uint32_t type = IREE_ELF_R_TYPE(rela->r_info);
+    if (type == 0) continue;
```

Comment:
Can you spell the '0' symbolically to make it more clear that you are early exiting for the same case covered below by the switch?